install_snap/module-artifact-gen/install_snap-artifact-gen

```diff
@@ -0,0 +1,138 @@
+#!/bin/bash
+
+set -e
+
+show_help() {
+  cat << EOF
+
+Simple tool to generate Mender Artifact suitable for install_snap Update Module
+
+Usage: $0 [options] SNAP [SNAPS...] [-- [options-for-mender-artifact] ]
+
+    Options: [ -n|artifact-name -t|--device-type -o|--output_path -h|--help ]
+
+        --artifact-name     - Artifact name
+        --device-type       - Target device type identification (can be given more than once)
+        --output-path       - Path to output file. Default: install_snap-artifact.mender
+        --help              - Show help and exit
+        SNAP                - snap name(s) to add to the Artifact
+
+Anything after a '--' gets passed directly to the mender-artifact tool.
+
+EOF
+}
+
+show_help_and_exit_error() {
+  show_help
+  exit 1
+}
+
+check_dependency() {
+  if ! which "$1" > /dev/null; then
+    echo "The $1 utility is not found but required to generate Artifacts." 1>&2
+    return 1
+  fi
+}
+
+if ! check_dependency mender-artifact; then
+  echo "Please follow the instructions here to install mender-artifact and then try again: https://docs.mender.io/downloads#mender-artifact" 1>&2
+  exit 1
+fi
+check_dependency snap
+check_dependency jq
+
+device_types=""
+artifact_name=""
+output_path="install_snap-artifact.mender"
+meta_data_file="meta-data.json"
+SNAPS=""
+passthrough=0
+passthrough_args=""
+
+while (( "$#" )); do
+  if test $passthrough -eq 1
+  then
+    passthrough_args="$passthrough_args $1"
+    shift
+    continue
+  fi
+  case "$1" in
+    --device-type | -t)
+      if [ -z "$2" ]; then
+        show_help_and_exit_error
+      fi
+      device_types="$device_types $1 $2"
+      shift 2
+      ;;
+    --artifact-name | -n)
+      if [ -z "$2" ]; then
+        show_help_and_exit_error
+      fi
+      artifact_name=$2
+      shift 2
+      ;;
+    --output-path | -o)
+      if [ -z "$2" ]; then
+        show_help_and_exit_error
+      fi
+      output_path=$2
+      shift 2
+      ;;
+    -h | --help)
+      show_help
+      exit 0
+      ;;
+    --)
+      passthrough=1
+      shift
+      ;;
+    -*)
+      echo "Error: unsupported option $1"
+      show_help_and_exit_error
+      ;;
+    *)
+      SNAPS="$SNAPS $1"
+      shift
+      ;;
+  esac
+done
+
+if [ -z "${artifact_name}" ]; then
+  echo "Artifact name not specified. Aborting."
+  show_help_and_exit_error
+fi
+
+if [ -z "${device_types}" ]; then
+  echo "Device type not specified. Aborting."
+  show_help_and_exit_error
+fi
+
+if [ -z "${SNAPS}" ]; then
+  echo "At least one snap must be specified. Aborting."
+  show_help_and_exit_error
+fi
+
+snaps=""
+for snap in $SNAPS; do
+  if ! snap info $snap >/dev/null 2>&1; then
+    echo "The ${snap} snap is not valid. Aborting."
+    show_help_and_exit_error
+  fi
+  snaps="$snaps \"${snap}\""
+done
+snaps=$(echo $snaps | tr ' ' ',')
```

Comment:
One possible future improvement would be to include the snaps in the software versioning somehow. Normally one uses the `--software-name` and `--software-version` options, but since you can install multiple at a time, you may need to use `--no-defaults-software-version`, and then specify them manually with `--provides rootfs-image.<SNAP_NAME>.version:SOMETHING`. The `SOMETHING` might also be a problem in the case of snap, I suppose.

Anyway, it's just a suggestion. See [software versioning](https://docs.mender.io/artifact-creation/software-versioning) for more details.